js/ui/handler/drag_rotate.js

```diff
@@ -16,74 +21,157 @@ function DragRotate(map) {
+    _onDown: function (e) {
+        if (this.active) return;
+        if (e.button !== 2) return;
+
+        document.addEventListener('mousemove', this._onMove);
+        document.addEventListener('mouseup', this._onUp);
+
+        this.active = false;
+        this._inertia = [[Date.now(), this._map.getBearing()]];
         this._startPos = this._pos = DOM.mousePos(this._el, e);
+        this._center = this._map.transform.centerPoint;  // Center of rotation
 
-        document.addEventListener('mousemove', this._onMouseMove, false);
-        document.addEventListener('mouseup', this._onMouseUp, false);
+        // If the first click was too close to the center, move the center of rotation by 200 pixels
+        // in the direction of the click.
+        var startToCenter = this._startPos.sub(this._center),
+            startToCenterDist = startToCenter.mag();
+
+        if (startToCenterDist < 200) {
+            this._center = this._startPos.add(new Point(-200, 0)._rotate(startToCenter.angle()));
+        }
 
         e.preventDefault();
     },
 
-    _onMouseMove: function (e) {
-
-        var p0 = this._startPos,
-            p1 = this._pos,
-            p2 = DOM.mousePos(this._el, e),
-
-            map = this._map,
-            center = map.transform.centerPoint, // Center of rotation
-            startToCenter = p0.sub(center),
-            startToCenterDist = startToCenter.mag();
+    _onMove: function (e) {
+        var map = this._map;
 
-        this.active = true;
+        if (e.buttons & 2 === 0) return;
+        if (map.boxZoom && map.boxZoom.active) return;
+        if (map.dragPan && map.dragPan.active) return;
 
-        if (!map.rotating) {
-            map.fire('movestart');
-            map.rotating = true;
+        if (!this.active) {
+            this.active = true;
+            this._fireEvent('rotatestart', e);
+            this._fireEvent('movestart', e);
         }
 
-        // If the first click was too close to the center, move the center of rotation by 200 pixels
-        // in the direction of the click.
-        if (startToCenterDist < 200) {
-            center = p0.add(new Point(-200, 0)._rotate(startToCenter.angle()));
-        }
+        map.stop();
 
-        var bearingDiff = p1.sub(center).angleWith(p2.sub(center)) / Math.PI * 180;
-        map.transform.bearing = map.getBearing() - bearingDiff;
+        var p1 = this._pos,
+            p2 = DOM.mousePos(this._el, e),
+            center = this._center,
+            bearingDiff = p1.sub(center).angleWith(p2.sub(center)) / Math.PI * 180,
+            bearing = map.getBearing() - bearingDiff,
+            inertia = this._inertia,
+            last = inertia[inertia.length - 1],
+            now = Date.now();
 
-        map.fire('move').fire('rotate');
+        map.transform.bearing = bearing;
+        inertia.push([now, map._normalizeBearing(bearing, last[1])]);
+        while (inertia.length > 1 && now - inertia[0][0] > 50) inertia.shift();
```

Comment:
Could we somehow cut out generic inertia logic into a separate file or utility method to DRY up inertia code between files?